From 996f0aeb7282f736f37c6e7e0ee1d2bbb1c9a580 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Fri, 17 May 2024 17:21:06 +0800
Subject: [PATCH 088/236] mtk: hostapd: get link channel information and
 synchronize to AP

'wpa_s->valid_links' is used to determine the connection is MLO or not,
and different ways are used to retrieve operating channel information.

Refactor center frequency calculation part in the function
uc_wpa_freq_info.
1. It does not have to set seg0 for 2GHz.
2. The original center frequency calculation is wrong for 2G. Also, center
   frequency 1 of BW 20/40 MHz can be derived from control frequency &
   secondary channel offset.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/ap/ucode.c         |  2 +-
 wpa_supplicant/ucode.c | 14 ++++++++++----
 2 files changed, 11 insertions(+), 5 deletions(-)

diff --git a/src/ap/ucode.c b/src/ap/ucode.c
index bbcc03084..16545cef9 100644
--- a/src/ap/ucode.c
+++ b/src/ap/ucode.c
@@ -662,7 +662,7 @@ out:
 	wpa_printf(MSG_INFO, "    * seg0: %d\n",
 			hostapd_get_oper_centr_freq_seg0_idx(conf));
 	wpa_printf(MSG_INFO, "    * seg1: %d\n",
-			hostapd_get_oper_centr_freq_seg0_idx(conf));
+			hostapd_get_oper_centr_freq_seg1_idx(conf));
 	wpa_printf(MSG_INFO, "    * oper_chwidth: %d\n",
 			hostapd_get_oper_chwidth(conf));
 
diff --git a/wpa_supplicant/ucode.c b/wpa_supplicant/ucode.c
index 2f92d61be..91e80eb20 100644
--- a/wpa_supplicant/ucode.c
+++ b/wpa_supplicant/ucode.c
@@ -310,10 +310,16 @@ uc_wpas_iface_status(uc_vm_t *vm, size_t nargs)
 		else if (is_6ghz_freq(bss->freq))
 			band_idx = 2;
 
-		wpa_drv_channel_info(wpa_s, &ci);
-		freq = ci.frequency;
-		center_freq1 = ci.center_frq1;
-		chwidth=ci.chanwidth;
+		if (wpa_s->valid_links) {
+			freq = wpa_s->links[link_id].freq;
+			center_freq1 = wpa_s->links[link_id].center_freq1;
+			chwidth = wpa_s->links[link_id].width;
+		} else {
+			wpa_drv_channel_info(wpa_s, &ci);
+			freq = ci.frequency;
+			center_freq1 = ci.center_frq1;
+			chwidth=ci.chanwidth;
+		}
 
 		if (bss->freq != center_freq1) {
 			if (is_24ghz)
-- 
2.45.2

